Restart(Source|Flow|Sink): Configurable stream restart deadline

r-glyde opened the pull request:
References #29291 (and sort of #29550)

Docs could probably also do with a bit of an update after this but was waiting to see general thoughts first.

Currently introducing additional overloaded methods (except for the deprecated javadsl) to try to keep binary compatibility but there are now quite a lot of these methods hanging around which might make this harder to follow. Another option could be to introduce a settings class for these sort of options. Probably wouldn't make much difference in this PR as we would still need keep existing methods and introduce new ones to receive the settings class but it might make future changes easier? Let me know what's preferred and happy to update.

---
Edit - dc3c1e7 includes creating `RestartSettings` to hold configuration which can then be passed as a whole to the actual restart methods. I think this ends up looking a bit tidier and expect it should be easier to add additional configuration options in the future if needed. I'm not sure what anyone thinks about exactly what config options should be part of this class but for now I've just put everything except `onlyOnFailures` (I quite like having the separate `withBackoff` and `onFailuresWithBackoff` functions).

Let me know if the original API is preferred and I can just remove that commit.

Commit:
add configurable reset deadline

* use resetDeadline over minBackoff
* scala and java overloads

Commit:
revert unnecessary scaladoc changes

Commit:
rename resetDeadline to maxRestartsWithin

Commit:
add  mima filters

laszlovandenhoek commented:
As to docs, see existing issue: #28178 

It appears @vaslabs is another poor soul running into this :)

Commit:
create RestartSettings class

* deprecate existing restart methods
* update tests to use new methods
* additional mima filter needed

patriknw commented on akka-stream/src/main/scala/akka/stream/RestartSettings.scala:
```diff
@@ -0,0 +1,74 @@
+  def withMaxRestarts(value: Int): RestartSettings = copy(maxRestarts = value)
+
+  /** Scala API: the duration after which to reset the restart count and current exponential backoff */
+  def withMaxRestartsWithin(value: FiniteDuration): RestartSettings = copy(maxRestartsWithin = value)
```
since maxRestarts and maxRestartsWithin are related and should be set together I'd suggest to only have

`def withMaxRestarts(count: Int, within: FiniteDuration)`

r-glyde replied:
After updating some of the docs, got a small question on this. Because this `maxRestartsWithin` is being used to determine when the backoff time resets back to `minBackoff` it isn't possible to configure this reset time without also setting `maxRestarts`? Is there a use case for wanting a separate `backoffResetAfter: FiniteDuration` which can be configured without needing to limit `maxRestarts`? That value could then be reused as what the `maxRestartsWithin` is doing? Or is this all a separate issue.

patriknw reviewed (commented):
looking great, one small suggestion...

r-glyde commented:
Thanks for taking a look @patriknw. I've addressed the comment on `withMaxRestarts`. Doesn't seem to be an issue with the new API using `RestartSettings` so I'll go ahead and start updating the docs around this as well.

Commit:
set maxRestarts and maxRestartsWithin from withMaxRestarts

patriknw reviewed (approved):
LGTM, thanks!

ennru commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala:
```diff
@@ -35,11 +33,14 @@ class RestartSpec extends StreamSpec(Map("akka.test.single-expect-default" -> "1
   private val minBackoff = 1.second.dilated
   private val maxBackoff = 3.seconds.dilated
 
+  private val shortRestartSettings = RestartSettings(shortMinBackoff, shortMaxBackoff, 0)
+  private val restartSettings = RestartSettings(minBackoff, maxBackoff, 0)
```
Yes! I've always thought this should go into something reusable/nameable.

ennru commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala:
```diff
@@ -290,6 +291,30 @@ class RestartSpec extends StreamSpec(Map("akka.test.single-expect-default" -> "1
       probe.cancel()
     }
+
+    "allow using maxRestartsWithin instead of minBackoff to determine maxRestarts reset time" in assertAllStagesStopped {
```
```suggestion
    "allow using withMaxRestarts instead of minBackoff to determine the maxRestarts reset time" in assertAllStagesStopped {
```

ennru commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/RestartSpec.scala:
```diff
@@ -491,20 +516,50 @@ class RestartSpec extends StreamSpec(Map("akka.test.single-expect-default" -> "1
       probe.sendComplete()
     }
+
+    "allow using maxRestartsWithin instead of minBackoff to determine maxRestarts reset time" in assertAllStagesStopped {
```
```suggestion
    "allow using withMaxRestarts instead of minBackoff to determine maxRestarts reset time" in assertAllStagesStopped {
```

ennru commented on akka-stream/src/main/scala/akka/stream/RestartSettings.scala:
```diff
@@ -0,0 +1,73 @@
+      maxRestarts: Int = maxRestarts,
+      maxRestartsWithin: FiniteDuration = maxRestartsWithin): RestartSettings =
+    new RestartSettings(minBackoff, maxBackoff, randomFactor, maxRestarts, maxRestartsWithin)
+
```
Adding a `toString` would be useful.

ennru reviewed (commented):
Thank you for working on this, it is an important usability improvement.
Just a tiny suggestion and I'm happy with it.

Commit:
update documentation for stream restarts

* document RestartSettings
* update operator docs
* reference BackoffOpts instead of Backoff

Commit:
adjust text for tests using withMaxRestarts

Commit:
add toString for RestartSettings

Commit:
format java docs

ennru reviewed (commented):
Re-read the descriptions and have a few more clarifying suggestions.

r-glyde commented on akka-docs/src/main/paradox/stream/operators/RestartSink/withBackoff.md:
```diff
@@ -6,19 +6,28 @@ Wrap the given @apidoc[Sink] with a @apidoc[Sink] that will restart it when it f
+* @ref:[RestartSource.withBackoff](../RestartSource/withBackoff.md)
+* @ref:[RestartSource.onFailuresWithBackoff](../RestartSource/onFailuresWithBackoff.md)
+* @ref:[RestartFlow.onFailuresWithBackoff](../RestartFlow/onFailuresWithBackoff.md)
+* @ref:[RestartFlow.withBackoff](../RestartFlow/withBackoff.md)
```
Ah, yeah! Sorry, went through updating the others and just forgot that this one didn't exist 🙂

Commit:
update docs following review

ennru reviewed (approved):
LGTM.

ennru commented:
Thank you so much for this work @r-glydem, highly appriciated!
